federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+	clientRetryInterval = 5 * time.Second
+
+	retryable    = true
+	notRetryable = false
+
+	doNotRetry = time.Duration(0)
+
+	UserAgentName = "Federation-Service-Controller"
+	KubeAPIQPS    = 20.0
+	KubeAPIBurst  = 30
+)
+
+type cachedService struct {
+
+	//release_1_3 client returns v1 service
+	lastState *v1.Service
+	// The state as successfully applied to the load balancer
+	appliedState *v1.Service
+
+	// Ensures only one goroutine can operate on this service at any given time.
+	mu sync.Mutex
+
+	// Controls error back-off
+	lastRetryDelay time.Duration
+}
+
+type serviceCache struct {
+	mu         sync.Mutex // protects serviceMap
+	serviceMap map[string]*cachedService
+}
+
+type ServiceController struct {
+	dns                dnsprovider.Interface
+	knownClusterSet    sets.String
+	//we need two clientsets to talk to federation apiserver now, federationClient talks to cluster
+	//kubeclient talks to services.
+	federationClient   federationclientset.Interface
+	federationName     string
+	clusterClientMap   map[string]clientset.Clientset
+	balancer           cloudprovider.LoadBalancer
+	zones              []dnsprovider.Zone
+	cache              *serviceCache
+	clusterClientCache *clusterClientCache
+	// A store of services, populated by the serviceController
+	serviceStore       cache.StoreToServiceLister
+	// Watches changes to all services
+	serviceController  *framework.Controller
+	// A store of services, populated by the serviceController
+	clusterStore       federationcache.StoreToClusterLister
+	// Watches changes to all services
+	clusterController  *framework.Controller
+	eventBroadcaster   record.EventBroadcaster
+	eventRecorder      record.EventRecorder
+	//clusterLister     federationcache.StoreToClusterLister
+	// services that need to be synced
+	queue              *workqueue.Type
+}
+
+// New returns a new service controller to keep cloud provider service resources
+// (like Kubernetes Services and DNS server records for service discovery) in sync with the registry.
+
+func New(federationClient federationclientset.Interface, dns dnsprovider.Interface, federationName string) *ServiceController {
+	// TODO: quinton:
+	//                 1. cloudprovider.Interface is used for DNS configuration?
+	//                 2. need one kubeClient per cluster, or look up based on type of cluster.
+	//                 3. Replace clustername with cluster api object reference.
+	broadcaster := record.NewBroadcaster()
+	//broadcaster.StartRecordingToSink(&unversioned_core.EventSinkImpl{Interface: kubeClient.Core().Events("")})
+	recorder := broadcaster.NewRecorder(api.EventSource{Component: "federated-service-controller"})
+
+	sc := &ServiceController{
+		dns:              dns,
+		knownClusterSet:  make(sets.String),
+		federationClient: federationClient,
+		clusterClientMap: make(map[string]clientset.Clientset),
+		federationName:   federationName,
+		cache:            &serviceCache{serviceMap: make(map[string]*cachedService)},
```

Comment:
This map is holds the service info from federation apiserver, cluster is not assigned yet, so far I did not see the need, I will add cluster info when necessary.